overlord/ifacestate/hotplug.go

```diff
@@ -30,3 +34,115 @@ func (m *InterfaceManager) HotplugDeviceAdded(devinfo *hotplug.HotplugDeviceInfo
 // HotplugDeviceRemoved gets called when a device is removed from the system.
 func (m *InterfaceManager) HotplugDeviceRemoved(devinfo *hotplug.HotplugDeviceInfo) {
 }
+
+// ensureUniqueName modifies proposedName so that it's unique according to isUnique predicate.
+// Uniqueness is achieved by appending a numeric suffix, or increasing existing suffix.
+func ensureUniqueName(proposedName string, isUnique func(string) bool) string {
+	// if the name is unique right away, do nothing
+	if isUnique(proposedName) {
+		return proposedName
+	}
+
+	// extract number suffix if present
+	pname := []rune(proposedName)
+	end := len(pname) - 1
+	suffixIndex := end
+	for suffixIndex >= 0 && unicode.IsDigit(pname[suffixIndex]) {
+		suffixIndex--
+	}
+
+	var suffixNumValue uint64
+	// if numeric suffix hasn't been found, append "-" before the number
+	if suffixIndex == end {
+		pname = append(pname, '-')
+		suffixIndex++
+	} else {
+		var err error
+		suffixNumValue, err = strconv.ParseUint(string(pname[suffixIndex+1:]), 10, 32)
+		if err != nil {
+			suffixIndex = end
+		}
+	}
+
+	// increase suffix value until we have a unique name
+	for {
+		suffixNumValue++
```

Comment:
There's no example how ensureUniqueName is combined with the other two, but won't we end up depending on order of registration and possibly a slot ending up with a different name after a reboot?

Reply:
These are low-level helpers  that will be used to generate names when we see a device for the first time. It can change only for slots that are not connected.  As soon as a slot gets connected (by the user, or by auto-connect where applicable), the name "sticks" and gets associated with given device by its *device key* (see #5759 for default device key computation), and it will be remembered in the state (and isUnique implementation in one of the followup PRs will make sure it's not taken by any other slot).